### DISCUSSION TITLE
Adding ShouldBeUnique support for notifications

### DISCUSSION THREAD

**Discussion by gbradley** [laravel/framework#38595] [Ideas]
Laravel 8.x supports unique queued jobs via the `ShouldBeUnique` interface. This can be used to avoid duplicating work under various scenarios.

Currently there is no similar support for queued notifications. Ideally, I'd like to be able to use the same `ShouldBeUnique` interface and define a `uniqueId()` method on the notification to provide the unique key. Instead, I have to create a queued job that sends the notification immediately, which seems overly complex.

**Comment by enes1004** [comment]
Same here.

Has tried ShouldBeUnique but not working.
Tried adding WithoutOverlapping middleware, not working (even though it is serialized inside the job, not working per assumption)

Even adding a job() method for specification that will be used when job of notification is constructed will be sufficient

I have worked around by using cache, which will be called on queue (I edited the original code, so I cannot guarantee there is no error, but it should give some insights)

```
namespace App\Notifications;

use Illuminate\Support\Facades\Cache;

trait UniqueNotification{
  /**
   * The unique ID of the job.
   *
   * @return string
   */
  public function shouldSend($notifiable, $channel){
    $this->initUnique();
    return !$this->already_sent;
  }

  public function initUnique(){
    $this->already_sent=false;
    if(Cache::get('notification_will_sent.'.static::class.".".$this->uniqueId())){
      $this->already_sent=true;
    }
    // get how many seconds should pass to allow duplicate to send (it is called in queue, so delay function of notification is not need to be considered.)
    $this->release_duplicate_timer=$this->release_duplicate_timer??60;
    Cache::put('notification_will_sent.'.static::class.".".$this->uniqueId(),true,now()->addSeconds($this->release_duplicate_timer));
  }


}

```
And on notification 

```

namespace App\Notifications;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;

class CommentUpdated  implements ShouldQueue
{
    use Queueable,UniqueNotification;
    /**
     * Create a new notification instance.
     *
     * @return void
     */
    public $resource;
    public function __construct($resource)
    {
     $this->resource=$resource;
     $this->initUnique();
    // $will_send=now()->ceilMinute(20);
     // $this->delay($will_send);
    }

  public function uniqueId()
  {
     return $this->resource->id;
  }

```

**Comment by heikokrebs** [comment]
Support this, ran into errors because i assumed this would work.

**Comment by peterjthomson** [comment]
Also be good to have this on the Mail class when using ShouldQueue. We seem to be able to add the ShouldBeUnique trait, but can't quire get it working on a Mail object.

**Comment by DarkcoderSe** [comment]
Yea, I think its much-needed thing right now.

**Comment by ChipNeedham** [comment]
Still needed. Would work great for debouncing notifications for actions that a user might do rapidly.

**Comment by ChipNeedham** [comment]
The bandaid I'm using now is wrapping my notification with a job. Works pretty nicely actually. Come holla if you want details

**Comment by MohammadZarifiyan** [comment]
I need this feature too.
